iroh/src/magicsock/transports/relay.rs

```diff
@@ -0,0 +1,423 @@
+use std::{
+    io,
+    task::{Context, Poll},
+};
+
+use bytes::Bytes;
+use iroh_base::{NodeId, RelayUrl};
+use n0_future::{
+    ready,
+    task::{self, AbortOnDropHandle},
+};
+use n0_watcher::{Watchable, Watcher as _};
+use smallvec::SmallVec;
+use tokio::sync::mpsc;
+use tokio_util::sync::PollSender;
+use tracing::{error, info_span, trace, warn, Instrument};
+
+use super::{Addr, Transmit};
+use crate::magicsock::RelayContents;
+
+mod actor;
+
+pub(crate) use self::actor::Config as RelayActorConfig;
+use self::actor::{RelayActor, RelayActorMessage, RelayRecvDatagram, RelaySendItem};
+
+#[derive(Debug)]
+pub(crate) struct RelayTransport {
+    /// Queue to receive datagrams from relays for [`quinn::AsyncUdpSocket::poll_recv`].
+    relay_datagram_recv_queue: mpsc::Receiver<RelayRecvDatagram>,
+    /// Channel on which to send datagrams via a relay server.
+    relay_datagram_send_channel: mpsc::Sender<RelaySendItem>,
+    actor_sender: mpsc::Sender<RelayActorMessage>,
+    _actor_handle: AbortOnDropHandle<()>,
+    my_relay: Watchable<Option<RelayUrl>>,
+    my_node_id: NodeId,
+}
+
+impl RelayTransport {
+    pub(crate) fn new(config: RelayActorConfig) -> Self {
+        let (relay_datagram_send_tx, relay_datagram_send_rx) = mpsc::channel(256);
+
+        let (relay_datagram_recv_tx, relay_datagram_recv_rx) = mpsc::channel(512);
```

Comment:
I know you're probably copying these values... but how did we end up with the recv queue being twice as long as the send queue?  Anyway, something to clean up some other time.